#include "mainwindow.h"
#include "ui_mainwindow.h"
#include <QVBoxLayout>
#include <QHBoxLayout>
#include <QComboBox>
#include <QSpinBox>
#include <QDialog>
#include <QLabel>
#include <QLineEdit>
#include <QPushButton>
#include <QMessageBox>
#include <QInputDialog>
#include <QSqlDatabase>
#include <QSqlQuery>
#include <QSqlError>
#include <QDebug>
#include <QListWidgetItem>

MainWindow::MainWindow(QWidget *parent)
    : QMainWindow(parent)
    , ui(new Ui::MainWindow)
    , currentBookId(-1)
{
    ui->setupUi(this);
    connect(ui->listBooks, &QListWidget::currentRowChanged,
            this, &MainWindow::on_listBooks_currentRowChanged);

    if (!connectToDatabase()) {
        return;
    }
    QSqlQuery testQuery;
    if (testQuery.exec("SELECT COUNT(*) as total, "
                      "SUM(CASE WHEN status='planned' THEN 1 ELSE 0 END) as planned, "
                      "SUM(CASE WHEN status='reading' THEN 1 ELSE 0 END) as reading, "
                      "SUM(CASE WHEN status='finished' THEN 1 ELSE 0 END) as finished "
                      "FROM books")) {
    } else {
        qDebug() << "–û—à–∏–±–∫–∞ –≤—ã–ø–æ–ª–Ω–µ–Ω–∏—è –∑–∞–ø—Ä–æ—Å–∞ —Å—Ç–∞—Ç–∏—Å—Ç–∏–∫–∏:" << testQuery.lastError().text();
    }

    loadBooksFromDB();
    loadShelvesList();
}
MainWindow::~MainWindow()
{
    if (db.isOpen()) {
        db.close();
    }
    delete ui;
}
// ============ –ü–û–î–ö–õ–Æ–ß–ï–ù–ò–ï –ö –ë–ê–ó–ï ============
bool MainWindow::connectToDatabase()
{

    db = QSqlDatabase::addDatabase("QPSQL");
    db.setHostName("localhost");
    db.setDatabaseName("bo1");
    db.setUserName("postgres");
    db.setPassword("12345");
    db.setPort(5432);

    if (!db.open()) {
        QString error = db.lastError().text();
        qDebug() << "–û—à–∏–±–∫–∞ –ø–æ–¥–∫–ª—é—á–µ–Ω–∏—è:" << error;
}
    qDebug() << "–£—Å–ø–µ—à–Ω–æ–µ –ø–æ–¥–∫–ª—é—á–µ–Ω–∏–µ –∫ PostgreSQL!";
    return true;
}
// ============ –ó–ê–ì–†–£–ó–ö–ê –ö–ù–ò–ì ============
void MainWindow::loadBooksFromDB()
{
    ui->listBooks->clear();

    QSqlQuery query;
    bool success = query.exec("SELECT id, title, author, status, rating FROM books ORDER BY title");

    qDebug() << "–ó–∞–ø—Ä–æ—Å –≤—ã–ø–æ–ª–Ω–µ–Ω?" << success;
    if (!success) {
        qDebug() << "–û—à–∏–±–∫–∞ –∑–∞–ø—Ä–æ—Å–∞:" << query.lastError().text();
        QMessageBox::warning(this, "–û—à–∏–±–∫–∞",
            "–ù–µ —É–¥–∞–ª–æ—Å—å –∑–∞–≥—Ä—É–∑–∏—Ç—å –∫–Ω–∏–≥–∏:\n" + query.lastError().text());
        return;
    }

    int count = 0;
    while (query.next()) {
        count++;
        int id = query.value(0).toInt();
        QString title = query.value(1).toString();
        QString author = query.value(2).toString();
        QString status = query.value(3).toString();
        int rating = query.value(4).toInt();

        QString bookInfo = title + " - " + author;

        QString statusIcon;
        if (status == "planned") statusIcon = "‚è≥";
        else if (status == "reading") statusIcon = "üìñ";
        else if (status == "finished") statusIcon = "‚úÖ";

        bookInfo += " " + statusIcon;

        if (rating > 0) {
            QString stars;
            for (int i = 0; i < rating; i++) stars += "‚òÖ";
            for (int i = rating; i < 5; i++) stars += "‚òÜ";
            bookInfo += " " + stars;
        }

        QListWidgetItem *item = new QListWidgetItem(bookInfo);
        item->setData(Qt::UserRole, id);

        if (status == "reading") {
            item->setBackground(QColor(232, 245, 233));
            item->setForeground(QColor(27, 94, 32));
        }
        else if (status == "finished") {
            item->setBackground(QColor(243, 229, 245));
            item->setForeground(QColor(74, 20, 140));
        }
        else if (status == "planned") {
            item->setBackground(QColor(232, 244, 252));
            item->setForeground(QColor(13, 71, 161));
        }
        ui->listBooks->addItem(item);
    }

    qDebug() << "–ó–∞–≥—Ä—É–∂–µ–Ω–æ –∫–Ω–∏–≥:" << count;

    for (int i = 0; i < ui->listBooks->count(); i++) {
        QListWidgetItem *item = ui->listBooks->item(i);
        int bookId = item->data(Qt::UserRole).toInt();

        QSqlQuery shelfQuery;
        shelfQuery.prepare(
            "SELECT s.name FROM shelves s "
            "JOIN book_shelves bs ON s.id = bs.shelf_id "
            "WHERE bs.book_id = ?"
        );
        shelfQuery.addBindValue(bookId);

        if (shelfQuery.exec()) {
            QStringList shelves;
            while (shelfQuery.next()) {
                shelves << shelfQuery.value(0).toString();
            }

            if (!shelves.isEmpty()) {
                QString currentText = item->text();
                currentText += " [" + shelves.join(", ") + "]";
                item->setText(currentText);
            }
        }
    }
    updateBookCount();
}
// ============ –û–ë–ù–û–í–õ–ï–ù–ò–ï –°–ß–Å–¢–ß–ò–ö–ê ============
void MainWindow::updateBookCount()
{
    QSqlQuery query;

    query.exec("SELECT COUNT(*) FROM books");
    query.next();
    int total = query.value(0).toInt();

    query.exec("SELECT status, COUNT(*) FROM books GROUP BY status");
    QString statusInfo;
    while (query.next()) {
        QString status = query.value(0).toString();
        int count = query.value(1).toInt();

        if (status == "planned") statusInfo += " ‚è≥:" + QString::number(count);
        else if (status == "reading") statusInfo += " üìñ:" + QString::number(count);
        else if (status == "finished") statusInfo += " ‚úÖ:" + QString::number(count);
    }

    ui->labelCount->setText("–í—Å–µ–≥–æ: " + QString::number(total) + " –∫–Ω–∏–≥" + statusInfo);
}

// ============ –ó–ê–ì–†–£–ó–ö–ê –î–ï–¢–ê–õ–ï–ô –ö–ù–ò–ì–ò ============
void MainWindow::loadBookDetails(int bookId)
{
    if (bookId == -1) {
        ui->labelDetails->setText("–í—ã–±–µ—Ä–∏—Ç–µ –∫–Ω–∏–≥—É –¥–ª—è –ø—Ä–æ—Å–º–æ—Ç—Ä–∞ –¥–µ—Ç–∞–ª–µ–π");
        return;
    }

    QSqlQuery query;
    query.prepare(
        "SELECT title, author, genre, year, pages, status, "
        "rating, notes, start_date, finish_date "
        "FROM books WHERE id = ?"
    );
    query.addBindValue(bookId);

    if (!query.exec() || !query.next()) {
        ui->labelDetails->setText("–ù–µ —É–¥–∞–ª–æ—Å—å –∑–∞–≥—Ä—É–∑–∏—Ç—å –∏–Ω—Ñ–æ—Ä–º–∞—Ü–∏—é –æ –∫–Ω–∏–≥–µ");
        return;
    }

    QString details = "<b>üìö " + query.value(0).toString() + "</b><br>";
    details += "‚úçÔ∏è –ê–≤—Ç–æ—Ä: " + query.value(1).toString() + "<br>";

    if (!query.value(2).isNull()) {
        details += "üè∑Ô∏è –ñ–∞–Ω—Ä: " + query.value(2).toString() + "<br>";
    }

    if (query.value(3).toInt() > 0) {
        details += "üìÖ –ì–æ–¥: " + query.value(3).toString() + "<br>";
    }

    if (query.value(4).toInt() > 0) {
        details += "üìÑ –°—Ç—Ä–∞–Ω–∏—Ü: " + query.value(4).toString() + "<br>";
    }

    QString status = query.value(5).toString();
    QString statusText;
    if (status == "planned") statusText = "‚è≥ –•–æ—á—É –ø—Ä–æ—á–∏—Ç–∞—Ç—å";
    else if (status == "reading") statusText = "üìñ –ß–∏—Ç–∞—é —Å–µ–π—á–∞—Å";
    else if (status == "finished") statusText = "‚úÖ –ü—Ä–æ—á–∏—Ç–∞–Ω–æ";
    details += "üìñ –°—Ç–∞—Ç—É—Å: " + statusText + "<br>";

    if (!query.value(6).isNull()) {
        int rating = query.value(6).toInt();
        QString stars;
        for (int i = 0; i < rating; i++) stars += "‚òÖ";
        details += "‚≠ê –û—Ü–µ–Ω–∫–∞: " + stars + " (" + QString::number(rating) + "/5)<br>";
    }

    if (!query.value(7).isNull() && !query.value(7).toString().isEmpty()) {
        details += "üìù –ó–∞–º–µ—Ç–∫–∏: " + query.value(7).toString() + "<br>";
    }

    ui->labelDetails->setText(details);
    currentBookId = bookId;
}

// ============ –ó–ê–ì–†–£–ó–ö–ê –°–ü–ò–°–ö–ê –ü–û–õ–û–ö ============
void MainWindow::loadShelvesList()
{
    ui->listShelves->clear();

    QSqlQuery query("SELECT id, name, description FROM shelves ORDER BY name");
    while (query.next()) {
        int id = query.value(0).toInt();
        QString name = query.value(1).toString();
        QString desc = query.value(2).toString();

        QString itemText = "üìö " + name;
        if (!desc.isEmpty()) {
            itemText += "\n   " + desc;
        }

        QListWidgetItem *item = new QListWidgetItem(itemText);
        item->setData(Qt::UserRole, id);
        ui->listShelves->addItem(item);
    }
}

// ============ –î–û–ë–ê–í–õ–ï–ù–ò–ï –ö–ù–ò–ì–ò (–ü–û–õ–ù–ê–Ø –í–ï–†–°–ò–Ø) ============
void MainWindow::on_btnAddBook_clicked()
{
    // –ü—Ä–æ—Å—Ç–æ–π –¥–∏–∞–ª–æ–≥ —Å —Ä–∞—Å—à–∏—Ä–µ–Ω–Ω—ã–º–∏ –ø–æ–ª—è–º–∏
    QDialog dialog(this);
    dialog.setWindowTitle("‚ûï –î–æ–±–∞–≤–∏—Ç—å –∫–Ω–∏–≥—É");
    dialog.setFixedSize(350, 350);

    QVBoxLayout *layout = new QVBoxLayout(&dialog);

    QLabel *labelTitle = new QLabel("–ù–∞–∑–≤–∞–Ω–∏–µ –∫–Ω–∏–≥–∏:*");
    QLineEdit *editTitle = new QLineEdit();
    layout->addWidget(labelTitle);
    layout->addWidget(editTitle);

    QLabel *labelAuthor = new QLabel("–ê–≤—Ç–æ—Ä:*");
    QLineEdit *editAuthor = new QLineEdit();
    layout->addWidget(labelAuthor);
    layout->addWidget(editAuthor);

    QLabel *labelGenre = new QLabel("–ñ–∞–Ω—Ä (–Ω–µ–æ–±—è–∑–∞—Ç–µ–ª—å–Ω–æ):");
    QLineEdit *editGenre = new QLineEdit();
    layout->addWidget(labelGenre);
    layout->addWidget(editGenre);

    QLabel *labelStatus = new QLabel("–°—Ç–∞—Ç—É—Å:");
    QComboBox *comboStatus = new QComboBox();
    comboStatus->addItem("‚è≥ –•–æ—á—É –ø—Ä–æ—á–∏—Ç–∞—Ç—å", "planned");
    comboStatus->addItem("üìñ –ß–∏—Ç–∞—é —Å–µ–π—á–∞—Å", "reading");
    comboStatus->addItem("‚úÖ –ü—Ä–æ—á–∏—Ç–∞–Ω–æ", "finished");
    layout->addWidget(labelStatus);
    layout->addWidget(comboStatus);

    QLabel *labelRating = new QLabel("–û—Ü–µ–Ω–∫–∞ (1-5):");
    QSpinBox *spinRating = new QSpinBox();
    spinRating->setRange(0, 5);
    spinRating->setSpecialValueText("–Ω–µ –æ—Ü–µ–Ω–µ–Ω–æ");
    spinRating->setValue(0);
    layout->addWidget(labelRating);
    layout->addWidget(spinRating);

    QHBoxLayout *buttonLayout = new QHBoxLayout();
    QPushButton *btnOk = new QPushButton("üíæ –°–æ—Ö—Ä–∞–Ω–∏—Ç—å");
    QPushButton *btnCancel = new QPushButton("‚ùå –û—Ç–º–µ–Ω–∞");

    btnOk->setStyleSheet("background-color: #28a745; color: white; padding: 8px; border-radius: 5px;");
    btnCancel->setStyleSheet("background-color: #dc3545; color: white; padding: 8px; border-radius: 5px;");

    buttonLayout->addStretch();
    buttonLayout->addWidget(btnOk);
    buttonLayout->addWidget(btnCancel);
    layout->addLayout(buttonLayout);

    // –û–±—Ä–∞–±–æ—Ç–∫–∞
    bool dialogAccepted = false;

    connect(btnOk, &QPushButton::clicked, [&]() {
        if (editTitle->text().trimmed().isEmpty()) {
            QMessageBox::warning(&dialog, "–û—à–∏–±–∫–∞", "–í–≤–µ–¥–∏—Ç–µ –Ω–∞–∑–≤–∞–Ω–∏–µ –∫–Ω–∏–≥–∏!");
            return;
        }
        if (editAuthor->text().trimmed().isEmpty()) {
            QMessageBox::warning(&dialog, "–û—à–∏–±–∫–∞", "–í–≤–µ–¥–∏—Ç–µ –∞–≤—Ç–æ—Ä–∞!");
            return;
        }
        dialogAccepted = true;
        dialog.accept();
    });

    connect(btnCancel, &QPushButton::clicked, &dialog, &QDialog::reject);

    if (dialog.exec() == QDialog::Accepted && dialogAccepted) {
        QSqlQuery query;
        query.prepare(
            "INSERT INTO books (title, author, genre, status, rating) "
            "VALUES (?, ?, ?, ?, ?)"
        );

        query.addBindValue(editTitle->text().trimmed());
        query.addBindValue(editAuthor->text().trimmed());

        QString genre = editGenre->text().trimmed();
        if (genre.isEmpty()) {
            query.addBindValue(QVariant());
        } else {
            query.addBindValue(genre);
        }

        // –°—Ç–∞—Ç—É—Å –∏ —Ä–µ–π—Ç–∏–Ω–≥
        query.addBindValue(comboStatus->currentData().toString());

        int rating = spinRating->value();
        if (rating > 0) {
            query.addBindValue(rating);
        } else {
            query.addBindValue(QVariant());
        }

        if (query.exec()) {
            // –ê–≤—Ç–æ–º–∞—Ç–∏—á–µ—Å–∫–∏ –¥–æ–±–∞–≤–ª—è–µ–º –Ω–∞ –ø–æ–ª–∫—É
            QString status = comboStatus->currentData().toString();
            int newBookId = query.lastInsertId().toInt();

            QSqlQuery shelfQuery;
            if (status == "planned") {
                shelfQuery.prepare("SELECT id FROM shelves WHERE name = '–•–æ—á—É –ø—Ä–æ—á–∏—Ç–∞—Ç—å'");
            } else if (status == "reading") {
                shelfQuery.prepare("SELECT id FROM shelves WHERE name = '–ß–∏—Ç–∞—é —Å–µ–π—á–∞—Å'");
            } else if (status == "finished") {
                shelfQuery.prepare("SELECT id FROM shelves WHERE name = '–ü—Ä–æ—á–∏—Ç–∞–Ω–æ'");
            }

            if (shelfQuery.exec() && shelfQuery.next()) {
                int shelfId = shelfQuery.value(0).toInt();

                QSqlQuery linkQuery;
                linkQuery.prepare("INSERT INTO book_shelves (book_id, shelf_id) VALUES (?, ?)");
                linkQuery.addBindValue(newBookId);
                linkQuery.addBindValue(shelfId);
                linkQuery.exec();
            }

            // –û–±–Ω–æ–≤–ª—è–µ–º –∏–Ω—Ç–µ—Ä—Ñ–µ–π—Å
            loadBooksFromDB();
            QMessageBox::information(this, "–£—Å–ø–µ—Ö", "–ö–Ω–∏–≥–∞ –¥–æ–±–∞–≤–ª–µ–Ω–∞! ‚ú®");
        } else {
            QMessageBox::warning(this, "–û—à–∏–±–∫–∞",
                "–ù–µ —É–¥–∞–ª–æ—Å—å –¥–æ–±–∞–≤–∏—Ç—å –∫–Ω–∏–≥—É:\n" + query.lastError().text());
        }
    }
}
// ============ –£–î–ê–õ–ï–ù–ò–ï –ö–ù–ò–ì–ò ============
void MainWindow::on_btnDelete_clicked()
{
    int row = ui->listBooks->currentRow();
    if (row == -1) {
        QMessageBox::warning(this, "–í–Ω–∏–º–∞–Ω–∏–µ", "–í—ã–±–µ—Ä–∏—Ç–µ –∫–Ω–∏–≥—É –¥–ª—è —É–¥–∞–ª–µ–Ω–∏—è!");
        return;
    }

    QListWidgetItem *item = ui->listBooks->item(row);
    int bookId = item->data(Qt::UserRole).toInt();
    QString bookTitle = item->text().split(" - ")[0];

    QMessageBox::StandardButton reply;
    reply = QMessageBox::question(this, "–ü–æ–¥—Ç–≤–µ—Ä–∂–¥–µ–Ω–∏–µ",
        "–£–¥–∞–ª–∏—Ç—å –∫–Ω–∏–≥—É –∏–∑ –±–∞–∑—ã –¥–∞–Ω–Ω—ã—Ö?\n\"" + bookTitle + "\"?\n\n"
        "–≠—Ç–æ —Ç–∞–∫–∂–µ —É–¥–∞–ª–∏—Ç –≤—Å–µ —Ü–∏—Ç–∞—Ç—ã –∏ —Å–≤—è–∑–∏ —Å –ø–æ–ª–∫–∞–º–∏.",
        QMessageBox::Yes | QMessageBox::No);

    if (reply == QMessageBox::Yes) {
        QSqlQuery query;
        query.prepare("DELETE FROM books WHERE id = ?");
        query.addBindValue(bookId);

        if (query.exec()) {
            loadBooksFromDB();
            ui->labelDetails->setText("–í—ã–±–µ—Ä–∏—Ç–µ –∫–Ω–∏–≥—É –¥–ª—è –ø—Ä–æ—Å–º–æ—Ç—Ä–∞ –¥–µ—Ç–∞–ª–µ–π");
            QMessageBox::information(this, "–£—Å–ø–µ—Ö", "–ö–Ω–∏–≥–∞ —É–¥–∞–ª–µ–Ω–∞!");
        } else {
            QMessageBox::warning(this, "–û—à–∏–±–∫–∞",
                "–ù–µ —É–¥–∞–ª–æ—Å—å —É–¥–∞–ª–∏—Ç—å –∫–Ω–∏–≥—É:\n" + query.lastError().text());
        }
    }
}

// ============ –í–´–ë–û–† –ö–ù–ò–ì–ò –í –°–ü–ò–°–ö–ï ============
void MainWindow::on_listBooks_currentRowChanged(int row)
{
    if (row == -1) {
        loadBookDetails(-1);
        return;
    }

    QListWidgetItem *item = ui->listBooks->item(row);
    int bookId = item->data(Qt::UserRole).toInt();
    loadBookDetails(bookId);
}

// ============ –î–û–ë–ê–í–õ–ï–ù–ò–ï –¶–ò–¢–ê–¢–´ ============
void MainWindow::on_btnAddQuote_clicked()
{
    if (currentBookId == -1) {
        QMessageBox::warning(this, "–û—à–∏–±–∫–∞", "–°–Ω–∞—á–∞–ª–∞ –≤—ã–±–µ—Ä–∏—Ç–µ –∫–Ω–∏–≥—É!");
        return;
    }

    bool ok;
    QString quote = QInputDialog::getText(this, "–ù–æ–≤–∞—è —Ü–∏—Ç–∞—Ç–∞",
        "–¢–µ–∫—Å—Ç —Ü–∏—Ç–∞—Ç—ã:", QLineEdit::Normal, "", &ok);
    if (!ok || quote.isEmpty()) return;

    int page = QInputDialog::getInt(this, "–ù–æ–≤–∞—è —Ü–∏—Ç–∞—Ç–∞",
        "–ù–æ–º–µ—Ä —Å—Ç—Ä–∞–Ω–∏—Ü—ã (0 –µ—Å–ª–∏ –Ω–µ —É–∫–∞–∑–∞–Ω):", 0, 0, 10000, 1, &ok);
    if (!ok) return;

    QSqlQuery query;
    query.prepare("INSERT INTO quotes (book_id, quote_text, page) VALUES (?, ?, ?)");
    query.addBindValue(currentBookId);
    query.addBindValue(quote);
    query.addBindValue(page);

    if (query.exec()) {
        QMessageBox::information(this, "–£—Å–ø–µ—Ö", "–¶–∏—Ç–∞—Ç–∞ –¥–æ–±–∞–≤–ª–µ–Ω–∞!");
    } else {
        QMessageBox::warning(this, "–û—à–∏–±–∫–∞",
            "–ù–µ —É–¥–∞–ª–æ—Å—å –¥–æ–±–∞–≤–∏—Ç—å —Ü–∏—Ç–∞—Ç—É:\n" + query.lastError().text());
    }
}

// ============ –ü–û–ö–ê–ó–ê–¢–¨ –¶–ò–¢–ê–¢–´ ============
void MainWindow::on_btnShowQuotes_clicked()
{
    if (currentBookId == -1) {
        QMessageBox::warning(this, "–û—à–∏–±–∫–∞", "–°–Ω–∞—á–∞–ª–∞ –≤—ã–±–µ—Ä–∏—Ç–µ –∫–Ω–∏–≥—É!");
        return;
    }

    QSqlQuery query;
    query.prepare("SELECT quote_text, page FROM quotes WHERE book_id = ? ORDER BY page");
    query.addBindValue(currentBookId);

    if (!query.exec()) {
        QMessageBox::warning(this, "–û—à–∏–±–∫–∞", "–ù–µ —É–¥–∞–ª–æ—Å—å –∑–∞–≥—Ä—É–∑–∏—Ç—å —Ü–∏—Ç–∞—Ç—ã");
        return;
    }

    QString quotesText = "<b>üìù –¶–∏—Ç–∞—Ç—ã –∏–∑ –∫–Ω–∏–≥–∏:</b><br><br>";
    bool hasQuotes = false;

    while (query.next()) {
        hasQuotes = true;
        QString quote = query.value(0).toString();
        int page = query.value(1).toInt();

        quotesText += "‚ùù" + quote + "‚ùû";
        if (page > 0) {
            quotesText += " (—Å—Ç—Ä. " + QString::number(page) + ")";
        }
        quotesText += "<br><br>";
    }

    if (!hasQuotes) {
        quotesText += "–ü–æ–∫–∞ –Ω–µ—Ç —Ü–∏—Ç–∞—Ç –∏–∑ —ç—Ç–æ–π –∫–Ω–∏–≥–∏";
    }

    QMessageBox::information(this, "–¶–∏—Ç–∞—Ç—ã", quotesText);
}

// ============ –î–û–ë–ê–í–ò–¢–¨ –ù–ê –ü–û–õ–ö–£ ============
void MainWindow::on_btnAddToShelf_clicked()
{
    if (currentBookId == -1) {
        QMessageBox::warning(this, "–û—à–∏–±–∫–∞", "–°–Ω–∞—á–∞–ª–∞ –≤—ã–±–µ—Ä–∏—Ç–µ –∫–Ω–∏–≥—É!");
        return;
    }

    int shelfRow = ui->listShelves->currentRow();
    if (shelfRow == -1) {
        QMessageBox::warning(this, "–û—à–∏–±–∫–∞", "–í—ã–±–µ—Ä–∏—Ç–µ –ø–æ–ª–∫—É!");
        return;
    }

    QListWidgetItem *shelfItem = ui->listShelves->item(shelfRow);
    int shelfId = shelfItem->data(Qt::UserRole).toInt();

    addToShelf(currentBookId, shelfId);
}

// ============ –ü–û–ú–ï–°–¢–ò–¢–¨ –ö–ù–ò–ì–£ –ù–ê –ü–û–õ–ö–£ ============
void MainWindow::addToShelf(int bookId, int shelfId)
{
    // –ü—Ä–æ–≤–µ—Ä—è–µ–º, –Ω–µ—Ç –ª–∏ —É–∂–µ —Ç–∞–∫–æ–π —Å–≤—è–∑–∏
    QSqlQuery checkQuery;
    checkQuery.prepare("SELECT 1 FROM book_shelves WHERE book_id = ? AND shelf_id = ?");
    checkQuery.addBindValue(bookId);
    checkQuery.addBindValue(shelfId);

    if (checkQuery.exec() && checkQuery.next()) {
        QMessageBox::information(this, "–ò–Ω—Ñ–æ—Ä–º–∞—Ü–∏—è",
            "–≠—Ç–∞ –∫–Ω–∏–≥–∞ —É–∂–µ –Ω–∞ —ç—Ç–æ–π –ø–æ–ª–∫–µ!");
        return;
    }

    // –î–æ–±–∞–≤–ª—è–µ–º —Å–≤—è–∑—å
    QSqlQuery query;
    query.prepare("INSERT INTO book_shelves (book_id, shelf_id) VALUES (?, ?)");
    query.addBindValue(bookId);
    query.addBindValue(shelfId);

    if (query.exec()) {
        // –û–±–Ω–æ–≤–ª—è–µ–º —Å–ø–∏—Å–æ–∫ –∫–Ω–∏–≥ —Å –Ω–æ–≤—ã–º–∏ –ø–æ–ª–∫–∞–º–∏
        loadBooksFromDB();

        // –ó–∞–≥—Ä—É–∂–∞–µ–º –¥–µ—Ç–∞–ª–∏ —Ç–µ–∫—É—â–µ–π –∫–Ω–∏–≥–∏ –∑–∞–Ω–æ–≤–æ
        if (currentBookId == bookId) {
            loadBookDetails(bookId);
        }

        QMessageBox::information(this, "–£—Å–ø–µ—Ö", "–ö–Ω–∏–≥–∞ –¥–æ–±–∞–≤–ª–µ–Ω–∞ –Ω–∞ –ø–æ–ª–∫—É!");
    } else {
        QMessageBox::warning(this, "–û—à–∏–±–∫–∞",
            "–ù–µ —É–¥–∞–ª–æ—Å—å –¥–æ–±–∞–≤–∏—Ç—å –Ω–∞ –ø–æ–ª–∫—É:\n" + query.lastError().text());
    }
}
// ============ –†–ï–î–ê–ö–¢–ò–†–û–í–ê–ù–ò–ï –ö–ù–ò–ì–ò ============
void MainWindow::on_btnEdit_clicked()
{
    int row = ui->listBooks->currentRow();
    if (row == -1) {
        QMessageBox::warning(this, "–í–Ω–∏–º–∞–Ω–∏–µ", "–í—ã–±–µ—Ä–∏—Ç–µ –∫–Ω–∏–≥—É –¥–ª—è —Ä–µ–¥–∞–∫—Ç–∏—Ä–æ–≤–∞–Ω–∏—è!");
        return;
    }

    QListWidgetItem *item = ui->listBooks->item(row);
    int bookId = item->data(Qt::UserRole).toInt();

    QSqlQuery query;
    query.prepare("SELECT title, author, status, rating FROM books WHERE id = ?");
    query.addBindValue(bookId);

    if (!query.exec() || !query.next()) {
        QMessageBox::warning(this, "–û—à–∏–±–∫–∞", "–ù–µ —É–¥–∞–ª–æ—Å—å –∑–∞–≥—Ä—É–∑–∏—Ç—å –¥–∞–Ω–Ω—ã–µ –∫–Ω–∏–≥–∏");
        return;
    }

    QString currentTitle = query.value(0).toString();
    QString currentAuthor = query.value(1).toString();
    QString currentStatus = query.value(2).toString();
    int currentRating = query.value(3).toInt();

    bool ok;
    QString newTitle = QInputDialog::getText(this, "–†–µ–¥–∞–∫—Ç–∏—Ä–æ–≤–∞—Ç—å",
        "–ù–∞–∑–≤–∞–Ω–∏–µ:", QLineEdit::Normal, currentTitle, &ok);
    if (!ok) return;

    QString newAuthor = QInputDialog::getText(this, "–†–µ–¥–∞–∫—Ç–∏—Ä–æ–≤–∞—Ç—å",
        "–ê–≤—Ç–æ—Ä:", QLineEdit::Normal, currentAuthor, &ok);
    if (!ok) return;

    QStringList statuses = {"planned", "reading", "finished"};
    QString currentStatusText;
    if (currentStatus == "planned") currentStatusText = "–•–æ—á—É –ø—Ä–æ—á–∏—Ç–∞—Ç—å";
    else if (currentStatus == "reading") currentStatusText = "–ß–∏—Ç–∞—é —Å–µ–π—á–∞—Å";
    else if (currentStatus == "finished") currentStatusText = "–ü—Ä–æ—á–∏—Ç–∞–Ω–æ";

    QString newStatusText = QInputDialog::getItem(this, "–†–µ–¥–∞–∫—Ç–∏—Ä–æ–≤–∞—Ç—å",
        "–°—Ç–∞—Ç—É—Å:", QStringList() << "–•–æ—á—É –ø—Ä–æ—á–∏—Ç–∞—Ç—å" << "–ß–∏—Ç–∞—é —Å–µ–π—á–∞—Å" << "–ü—Ä–æ—á–∏—Ç–∞–Ω–æ",
        statuses.indexOf(currentStatus), false, &ok);
    if (!ok) return;

    int newRating = QInputDialog::getInt(this, "–†–µ–¥–∞–∫—Ç–∏—Ä–æ–≤–∞—Ç—å",
        "–û—Ü–µ–Ω–∫–∞ (0-5, 0 - –Ω–µ –æ—Ü–µ–Ω–µ–Ω–æ):", currentRating, 0, 5, 1, &ok);
    if (!ok) return;

    QString newStatus;
    if (newStatusText == "–•–æ—á—É –ø—Ä–æ—á–∏—Ç–∞—Ç—å") newStatus = "planned";
    else if (newStatusText == "–ß–∏—Ç–∞—é —Å–µ–π—á–∞—Å") newStatus = "reading";
    else if (newStatusText == "–ü—Ä–æ—á–∏—Ç–∞–Ω–æ") newStatus = "finished";

    QSqlQuery updateQuery;
    updateQuery.prepare("UPDATE books SET title = ?, author = ?, status = ?, rating = ? WHERE id = ?");
    updateQuery.addBindValue(newTitle);
    updateQuery.addBindValue(newAuthor);
    updateQuery.addBindValue(newStatus);
    updateQuery.addBindValue(newRating > 0 ? newRating : QVariant());
    updateQuery.addBindValue(bookId);

    if (updateQuery.exec()) {
        loadBooksFromDB();
        loadBookDetails(bookId);
        QMessageBox::information(this, "–£—Å–ø–µ—Ö", "–ö–Ω–∏–≥–∞ –æ–±–Ω–æ–≤–ª–µ–Ω–∞!");
    }
}
